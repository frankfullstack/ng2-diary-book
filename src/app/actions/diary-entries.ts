import { Action } from '@ngrx/store';

import { DiaryEntrySet } from '../models/diary-entry-set';

export const LOAD_LIST         = '[Diary entries] Load list';
export const LOAD_LIST_SUCCESS = '[Diary entries] Load list success';
export const LOAD_LIST_FAILURE = '[Diary entries] Load list failure';

export const CREATE_ENTRY         = '[Diary entries] Create entry';
export const CREATE_ENTRY_SUCCESS = '[Diary entries] Create entry success';
export const CREATE_ENTRY_FAILURE = '[Diary entries] Create entry failure';


/**
 * Every action is comprised of at least a type and an optional
 * payload. Expressing actions as classes enables powerful
 * type checking in reducer functions.
 *
 * See Discriminated Unions: https://www.typescriptlang.org/docs/handbook/advanced-types.html#discriminated-unions
 */
export class LoadListAction implements Action {
  readonly type = LOAD_LIST;
}

export class LoadListSuccessAction implements Action {
  readonly type = LOAD_LIST_SUCCESS;

  constructor(public payload: object[]) { }
}

export class LoadListFailureAction implements Action {
  readonly type = LOAD_LIST_FAILURE;

  constructor(public payload: object) { }
}

export class CreateEntryAction implements Action {
  readonly type = CREATE_ENTRY;

  constructor(public payload: DiaryEntrySet) { }
}

export class CreateEntrySuccessAction implements Action {
  readonly type = CREATE_ENTRY_SUCCESS;

  constructor(public payload: any) { }
}

export class CreateEntryFailureAction implements Action {
  readonly type = CREATE_ENTRY_FAILURE;

  constructor(public payload: any) { }
}

/**
 * Export a type alias of all actions in this action group
 * so that reducers can easily compose action types
 */
export type Actions
  = LoadListAction
  | LoadListSuccessAction
  | CreateEntryFailureAction
  | CreateEntrySuccessAction
  | CreateEntryAction
  | LoadListFailureAction;
