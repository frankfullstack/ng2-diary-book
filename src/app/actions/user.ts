import { Action } from '@ngrx/store';

export const LOGIN    =  '[User] Login';
export const LOGOUT  =  '[User] Logout';


/**
 * Every action is comprised of at least a type and an optional
 * payload. Expressing actions as classes enables powerful
 * type checking in reducer functions.
 *
 * See Discriminated Unions: https://www.typescriptlang.org/docs/handbook/advanced-types.html#discriminated-unions
 */
export class LoginAction implements Action {
  readonly type = LOGIN;

  constructor(public payload: any) { }
}

export class LogoutAction implements Action {
  readonly type = LOGOUT;
}

/**
 * Export a type alias of all actions in this action group
 * so that reducers can easily compose action types
 */
export type Actions
  = LogoutAction
  | LoginAction;
