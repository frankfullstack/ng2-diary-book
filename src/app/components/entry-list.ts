import { Component, Input } from '@angular/core';
import { DiaryEntry } from '../models/diary-entry';

@Component({
  selector: 'diary-entry-list',
  template: `
    <div>
      <div *ngFor="let entry of entries">
        <diary-list-item [entry]="entry"></diary-list-item>
      </div>
    </div>
  `
})

export class EntryListComponent {
  @Input() entries: DiaryEntry[];

  constructor() {
  }
}
