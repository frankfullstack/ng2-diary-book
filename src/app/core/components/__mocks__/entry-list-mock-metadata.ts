const EntryListMockMetadata = {
  template: 'entry-list',
};

export { EntryListMockMetadata };
