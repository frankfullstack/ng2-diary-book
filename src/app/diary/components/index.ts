import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { MaterialModule, MdDatepickerModule, MdNativeDateModule } from '@angular/material';
import { ReactiveFormsModule, FormsModule } from '@angular/forms';
import { RouterModule } from '@angular/router';

import { DiaryComponent } from './diary';
import { EntryDetailComponent } from './entry-detail';
import { CommonDatePickerComponent } from './common-date-picker';

import { PipesModule } from '../../shared/pipes';

export const COMPONENTS = [
  DiaryComponent,
  EntryDetailComponent,
  CommonDatePickerComponent,
];

@NgModule({
  imports: [
    CommonModule,
    ReactiveFormsModule,
    MaterialModule,
    RouterModule,
    FormsModule,
    PipesModule,
    MdNativeDateModule,
    MdDatepickerModule,
    MdDatepickerModule,
  ],
  declarations: COMPONENTS,
  exports: COMPONENTS,
})
export class ComponentsModule {}
