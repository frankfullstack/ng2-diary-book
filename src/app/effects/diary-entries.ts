import 'rxjs/add/operator/catch';
import 'rxjs/add/operator/map';
import 'rxjs/add/operator/switchMap';
import 'rxjs/add/operator/filter';
import 'rxjs/add/operator/withLatestFrom';
import 'rxjs/add/operator/debounceTime';
import 'rxjs/add/operator/skip';
import 'rxjs/add/operator/takeUntil';
import { Injectable } from '@angular/core';
import { Effect, Actions, toPayload } from '@ngrx/effects';
import {Action, Store } from '@ngrx/store';
import { Observable } from 'rxjs/Observable';
import { empty } from 'rxjs/observable/empty';
import { of } from 'rxjs/observable/of';

import * as fromRoot from '../reducers';

import { DiaryEntryService } from '../services/diary-entry';
import * as diaryEntry from '../actions/diary-entries';

import { DiaryEntry } from '../models/diary-entry';
import { DiaryEntrySet } from '../models/diary-entry-set';

import * as moment from 'moment';

/**
 * Effects offer a way to isolate and easily test side-effects within your
 * application.
 * The `toPayload` helper function returns just
 * the payload of the currently dispatched action, useful in
 * instances where the current state is not necessary.
 *
 * Documentation on `toPayload` can be found here:
 * https://github.com/ngrx/effects/blob/master/docs/api.md#topayload
 *
 * If you are unfamiliar with the operators being used in these examples, please
 * check out the sources below:
 *
 * Official Docs: http://reactivex.io/rxjs/manual/overview.html#categories-of-operators
 * RxJS 5 Operators By Example: https://gist.github.com/btroncone/d6cf141d6f2c00dc6b35
 */

@Injectable()
export class DiaryEntriesEffects {
  @Effect()
  load$: Observable<Action> = this.actions$
    .ofType(diaryEntry.LOAD_LIST)
    .withLatestFrom(this.store)
    .filter(([action, state]) => fromRoot.getIsLoggedIn(state))
    .switchMap(([action, state]) => {
      const uid = fromRoot.getUid(state);

      return this.diaryEntryService.retrieveEntries(uid)
        .map((diaryEntries: DiaryEntry[]) => new diaryEntry.LoadListSuccessAction(diaryEntries))
        .catch(() => of(new diaryEntry.LoadListFailureAction([])));
    });

  @Effect()
  create$: Observable<Action> = this.actions$
    .ofType(diaryEntry.CREATE_ENTRY)
    .withLatestFrom(this.store)
    .switchMap(([action, state]) => {
      const uid = fromRoot.getUid(state);

      return this.diaryEntryService.createEntry(uid, action.payload)
        .map((newEntryData: DiaryEntry) => new diaryEntry.CreateEntrySuccessAction(newEntryData))
        .catch(() => of(new diaryEntry.CreateEntryFailureAction([])));
    });

  @Effect()
  edit$: Observable<Action> = this.actions$
    .ofType(diaryEntry.EDIT_ENTRY)
    .withLatestFrom(this.store)
    .switchMap(([action, state]) => {
      const uid = fromRoot.getUid(state);

      return this.diaryEntryService.updateEntry(uid, action.payload)
        .map((updatedEntryData: DiaryEntry) =>
          new diaryEntry.EditEntrySuccessAction(updatedEntryData)
        )
        .do(() => this.store.dispatch(new diaryEntry.LoadListAction()))
        .catch(() => of(new diaryEntry.EditEntryFailureAction(null)));
    });

    constructor(
      private actions$: Actions,
      private diaryEntryService: DiaryEntryService,
      private store: Store<fromRoot.State>,
    ) { }
}
