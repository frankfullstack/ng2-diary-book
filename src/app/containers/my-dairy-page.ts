import { Component, ChangeDetectorRef } from '@angular/core';
import { Observable } from 'rxjs/Observable';
import { Store } from '@ngrx/store';

import * as fromRoot from '../reducers/index';
import {isLoggedIn} from '../reducers/user';

import * as diaryEntries from '../actions/diary-entries';

import { DiaryEntryService } from '../services/diary-entry';
import { DiaryEntry } from '../models/diary-entry';

@Component({
  selector: 'diary-page',
  template: `
    <div [ngSwitch]="isLoggedIn">
      <div *ngSwitchCase="true">my dairy page</div>
      <div>
        <diary-entry-create></diary-entry-create>
      </div>
      <div>
        <diary-entry-list [entries]="diaryEntries$ | async"></diary-entry-list>
      </div>
      <div *ngSwitchCase="false">
        <diary-not-logged-in></diary-not-logged-in>
      </div>
    </div>
  `
})
export class MyDairyPageComponent {
  isLoggedIn: boolean;
  isLoggedIn$: Observable<boolean>;
  diaryEntries$: Observable<DiaryEntry[]>;

  constructor(
    private store: Store<fromRoot.State>,
    private cd: ChangeDetectorRef
  ) {
    this.isLoggedIn$ = store.select(fromRoot.getIsLoggedIn);
    this.diaryEntries$ = store.select(fromRoot.getEntries);
    this.isLoggedIn$
      .subscribe((newIsLoggedIn: boolean) => {
        this.isLoggedIn = newIsLoggedIn;
        this.cd.markForCheck();
      });

    this.store.dispatch(new diaryEntries.LoadListAction());
  }
}
