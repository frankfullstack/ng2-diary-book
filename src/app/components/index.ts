import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { MaterialModule } from '@angular/material';
import { MdDatepickerModule, MdNativeDateModule } from '@angular/material';
import { ReactiveFormsModule } from '@angular/forms';
import { RouterModule } from '@angular/router';
import { FormsModule } from '@angular/forms';

import { BookAuthorsComponent } from './book-authors';
import { BookDetailComponent } from './book-detail';
import { BookPreviewComponent } from './book-preview';
import { BookPreviewListComponent } from './book-preview-list';
import { BookSearchComponent } from './book-search';
import { LayoutComponent } from './layout';
import { NavItemComponent } from './nav-item';
import { SidenavComponent } from './sidenav';
import { ToolbarComponent } from './toolbar';
import { NotLoggedInComponent } from './not-logged-in';
import { EntryDetailComponent } from './entry-detail';
import { CommonDatePickerComponent } from './common-date-picker';
import { EntryListItemComponent } from './entry-list-item';
import { EntryListComponent } from './entry-list';

import { PipesModule } from '../pipes';


export const COMPONENTS = [
  BookAuthorsComponent,
  BookDetailComponent,
  BookPreviewComponent,
  BookPreviewListComponent,
  BookSearchComponent,
  LayoutComponent,
  NavItemComponent,
  SidenavComponent,
  ToolbarComponent,
  EntryDetailComponent,
  NotLoggedInComponent,
  CommonDatePickerComponent,
  EntryListComponent,
  EntryListItemComponent,
];


@NgModule({
  imports: [
    CommonModule,
    ReactiveFormsModule,
    MaterialModule,
    RouterModule,
    PipesModule,
    FormsModule,
    MdNativeDateModule,
    MdDatepickerModule,
  ],
  declarations: COMPONENTS,
  exports: COMPONENTS
})
export class ComponentsModule { }
