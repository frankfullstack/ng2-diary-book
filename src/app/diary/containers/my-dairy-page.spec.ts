/* tslint:disable:no-unused-variable */
import { async, ComponentFixture, TestBed } from '@angular/core/testing';

import { Store, StoreModule, combineReducers } from '@ngrx/store';

import * as fromFeature from '../reducers';
import * as fromRoot from '../../reducers';

import { MyDairyPageComponent } from './my-dairy-page';
import { ComponentsModule } from '../components';
import { CoreModule } from '../../core/core.module';
import { DialogFactoryService } from '../services/dialog-factory';
import { DiaryProcessorService } from '../services/diary-processor';

import { MdIconModule } from '@angular/material';

import * as userActions from '../../core/actions/user';
import * as diaryActions from '../actions/diary-entries';

import { EntryListMockMetadata } from '../../core/components/__mocks__/entry-list-mock-metadata';
import { EntryListComponent } from '../../core/components/entry-list';

describe('MyDairyPageComponent', () => {
  let component: MyDairyPageComponent;
  let fixture: ComponentFixture<MyDairyPageComponent>;
  let store: Store<fromFeature.State>;

  beforeEach(
    async(() => {
      TestBed.overrideComponent(EntryListComponent, {
        set: EntryListMockMetadata,
      })
        .configureTestingModule({
          imports: [
            MdIconModule,
            StoreModule.forRoot({
              ...fromRoot.reducers,
              diary: combineReducers(fromFeature.reducers),
            }),
            ComponentsModule,
            CoreModule,
          ],
          declarations: [MyDairyPageComponent],
          providers: [DialogFactoryService, DiaryProcessorService],
        })
        .compileComponents();

      store = TestBed.get(Store);
    }),
  );

  beforeEach(() => {
    fixture = TestBed.createComponent(MyDairyPageComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should match snapshot', () => {
    expect(fixture).toMatchSnapshot();
  });

  describe('полная имитация', () => {
    it('', () => {
      const userDataMock = {
        /*...*/
      };
      store.dispatch(new userActions.LoadUserAction(userDataMock));
      store.dispatch(
        new diaryActions.LoadListSuccessAction([
          {
            createDate: '2017-08-19T23:22:34.856Z',
            date: '2017-08-19T21:00:00.000Z',
            message: '13123123asdasdasd',
            $key: '-KrwnXF5o7VRJDTXJw-I',
          },
          {
            createDate: '2017-08-17T15:24:17.399Z',
            date: '2017-08-16T21:00:00.000Z',
            message: '123123123123123',
            $key: '-KrkmsupthRsqGXqVBeK',
          },
        ]),
      );

      fixture.detectChanges();

      expect(fixture).toMatchSnapshot();
    });
  });
});
