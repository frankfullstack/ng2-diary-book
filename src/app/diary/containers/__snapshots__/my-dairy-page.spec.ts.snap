// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyDairyPageComponent should match snapshot 1`] = `
"
    <common-show-if-logged-in _ngcontent-c4=\\"\\">
    <div ng-reflect-ng-switch=\\"false\\">
      <!--bindings={
  \\"ng-reflect-ng-switch-case\\": \\"true\\"
}-->
      <!--bindings={
  \\"ng-reflect-ng-switch-case\\": \\"false\\"
}--><div>
        <diary-not-logged-in>
    You are not logged in. Click to log in.
  </diary-not-logged-in>
      </div>
    </div>
  </common-show-if-logged-in>
  "
`;
